File: app/src/ultimate/java/com/clockmods/ui/WidgetWeatherBitmap.java
package com.clockmods.ui;
import android.content.Context;
import android.graphics.*;
/** Bounded bitmap facade; the parser and icon assets stay shared with WeatherIconView. */
public final class WidgetWeatherBitmap {
    private WidgetWeatherBitmap() { }
    public static Bitmap render(Context c,String code,boolean fill,int color,int size) {
        WeatherIcon icon=WeatherIcon.load(c,code,fill);
        if(icon==null) icon=WeatherIcon.load(c,"999",fill);
        Bitmap bitmap=Bitmap.createBitmap(size,size,Bitmap.Config.ARGB_8888);
        Paint p=new Paint(Paint.ANTI_ALIAS_FLAG); p.setColor(color);
        Canvas canvas=new Canvas(bitmap);
        if(icon!=null) icon.draw(canvas,0,0,size,p);
        else { p.setStyle(Paint.Style.STROKE); p.setStrokeWidth(size/12f); canvas.drawCircle(size/2f,size/2f,size/3f,p); }
        return bitmap;
    }
}

File: app/src/ultimate/java/com/clockmods/ui/WidgetWeatherBitmap.kt
package com.clockmods.ui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint

/** Bounded bitmap facade shared by RemoteViews widgets and the legacy weather view. */
object WidgetWeatherBitmap {
    @JvmStatic
    fun render(context: Context, code: String?, fill: Boolean, color: Int, size: Int): Bitmap {
        val boundedSize = size.coerceAtLeast(1)
        val icon = WeatherIcon.load(context, code, fill) ?: WeatherIcon.load(context, "999", fill)
        val bitmap = Bitmap.createBitmap(boundedSize, boundedSize, Bitmap.Config.ARGB_8888)
        val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { this.color = color }
        val canvas = Canvas(bitmap)
        if (icon != null) {
            icon.draw(canvas, 0f, 0f, boundedSize.toFloat(), paint)
        } else {
            paint.style = Paint.Style.STROKE
            paint.strokeWidth = boundedSize / 12f
            canvas.drawCircle(boundedSize / 2f, boundedSize / 2f, boundedSize / 3f, paint)
        }
        return bitmap
    }
}
